### PR TITLE
[Update 8] - Add custom repository support

### PR DESCRIPTION
**Pull request by Thevakumar-Luheerathan**
## Checklist

- [ ] **Page addition**
  - [ ] Add `permalink` to pages.

- [ ] **Page removal**
  - [ ] Remove entry from corresponding left nav YAML file.
  - [ ] Add `redirect_from` on the alternative page.
  - [ ] If no alternative page, add redirection on the `redirections.js ` file.

- [ ] **Page rename**
  - [ ] Add front-matter `redirect_from`.
  - [ ] Add front-matter `redirect_to:` (if applicable).

- [ ] **Page restrcuture**
  - [ ] Add `permalink` to pages.
  - [ ] Add front-matter `redirect_from`.
  - [ ] Add front-matter `redirect_to:` (if applicable).

### CODE REVIEW

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ A package can depend on other packages that are available in Ballerina repositor
 * The distribution repository
 * The Ballerina Central repository
 
-It also supports a third repository named the `local repository`. It temporarily overrides dependencies, which is useful for the package development and bug fixing phases.
+It also supports a third repository named the `local repository`. It temporarily overrides dependencies, which is useful for the package development and bug fixing phases. Additionally Maven custom repositories are also supported, which are useful to bring the user's private repositories into the dependency management.
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:17]:
```suggestion
It also supports a third repository named the `local repository`. It temporarily overrides dependencies, which is useful for the package development and bug fixing phases. Additionally, a predefined set of custom package repositories are also supported, which are useful to bring third-party repositories into dependency management.
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 2]
```diff
@@ -29,6 +29,10 @@ The Ballerina Central is a remote repository and creates a local file system cac
 The local repository is also a file system repository, which will be created in the `<USER_HOME>` location. The repository location is `<USER_HOME>/.ballerina/repositories/local/bala`. 
 For more information, see [Use dependencies from the local repository](/learn/manage-dependencies/#use-dependencies-from-the-local-repository).
 
+**Maven custom repositories**
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:32]:
```suggestion
**Custom repositories**
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 3]
```diff
@@ -29,6 +29,10 @@ The Ballerina Central is a remote repository and creates a local file system cac
...
+
+Maven custom are remote repositories and create a local file system cache at `<USER_HOME>/.ballerina/repositories/<REPOSITORY_ID>/bala`. Ballerina queries the remote repository only if the specified dependency version is not present in its local cache. For more information, see [Use Maven repositories for the dependency management](/learn/manage-dependencies/#use-maven-repositories-for-the-dependency-management).
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:34]:
```suggestion
Custom repositories

Ballerina supports one or more custom remote repositories which can be configured in the `<USER_HOME>/.ballerina/Settings.toml`. A local filesystem cache is maintained per repository at `<USER_HOME>/.ballerina/repositories/<REPOSITORY_ID>/bala`. Ballerina queries the remote repository only if the specified dependency version is not present in its local cache. For more information, see [Use dependencies from a custom repository](/learn/manage-dependencies/#use-dependencies-from-a-custom-repository).
```

Correct the link if necessary.

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 4]
```diff
@@ -153,6 +157,57 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
 However, the compiler gives priority to the latest version if a new patch version is found in distribution or Ballerina Central repositories.
 At this point, the compiler resolves the latest version and ignores the dependency version in the local repository.
 
+
+## Use Maven repositories for the dependency management
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:161]:
```suggestion
## Use dependencies from a custom repository
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 5]
```diff
@@ -153,6 +157,57 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+
+Currently, [Nexus](https://www.sonatype.com/products/sonatype-nexus-repository), [Artifactory](https://jfrog.com/artifactory/) and [Github packages](https://github.com/features/packages) are only supported. To use this feature `Settings.toml` needs to be configured with the proper repository configurations.
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:163]: Ballerina supports [Nexus](https://www.sonatype.com/products/sonatype-nexus-repository), [Artifactory](https://jfrog.com/artifactory/) and [Github packages](https://github.com/features/packages) to be set up as custom repositories. You can configure the custom repository in `<USER_HOME>/.ballerina/Settings.toml` to integrate it to package resolution.

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 6]
```diff
@@ -153,6 +157,57 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+
+```toml
+[[repository.TYPE]]
+id = "<repository-id>" # This id is used when pushing/ pulling balas
+url = "<repository-url>"
+userId = "<username/ specific-userId>"
+accesstoken = "<password/ accesstoken>"
+```
+Here `TYPE` can be anything among `nexus`, `artifactory`, `github`.
+
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:173]:
```suggestion

The following example shows how to configure a GitHub package repository to resolve a specific dependency.

```
Add an example of the configuration in Settings.toml and the Ballerina.toml

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 7]
```diff
@@ -153,6 +157,57 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+### Publish Balllerina archieve
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:174]:
```suggestion
### Publish the Balllerina archive (.bala)
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 8]
```diff
@@ -153,6 +157,57 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+   $ bal push --repository <repo-id>
+   ```
+
+   If you already have the path of Ballerina archive, then,
+
+    ```
+    $ bal push --repository <repo-id> <path-to-bala-archive>
+    ```
+
+### Pull Balllerina archieve
```
[the remark this answers is not in the document]
Comment (azinneera) [reply swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:192]: Pull the Ballerina library package

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 9]
```diff
@@ -153,6 +157,57 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+
+1. Ballerina package from the custom remote repository can be pulled using following command.
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:194]:
```suggestion
1. Ballerina package from the custom remote repository can be pulled using the following command. 
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 10]
```diff
@@ -153,6 +157,57 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+    ```
+
+### Pull Balllerina archieve
+
+1. Ballerina package from the custom remote repository can be pulled using following command. 
+    ```
+   $ bal pull <org>/<package-name>:<version> --repository <repo-id>
+    ```
+
+### Use the dependency to build package
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:199]:
```suggestion
### Use a dependency from a custom repository
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md (hunk 6 above)
Comment (azinneera) [reply swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:173]: Add an example of a GitHub packages configuration. Use github as the repo-id

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 11]
```diff
@@ -153,6 +157,57 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+
+1. Similar to `local` repository, custom repository packages can be specified in the `Ballerina.toml` file and utilised in a similar manner.
+
+    ```toml
```
[the remark this answers is not in the document]
Comment (azinneera) [reply swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:203]: Use the repo details shown in the Github example to show the dependency.

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 12]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+
+## Use dependencies from a custom repository
+
+Ballerina supports [Nexus](https://www.sonatype.com/products/sonatype-nexus-repository), [Artifactory](https://jfrog.com/artifactory/) and [Github packages](https://github.com/features/packages) to be set up as custom repositories. You can configure the custom repository in `<USER_HOME>/.ballerina/Settings.toml` to integrate it to package resolution.
+
+```toml
+[[repository.TYPE]]
+id = "<repository-id>" # This id is used when pushing/ pulling balas
+url = "<repository-url>"
+userId = "<username/ specific-userId>"
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:169]:
```suggestion
username = "<username/userId>"
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 13]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+accesstoken = "<password/ accesstoken>"
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:170]:
```suggestion
accesstoken = "<password/accesstoken>"
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 14]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+```
+Here `TYPE` can be anything among `nexus`,  `artifactory`,  `github`. The following steps show how to configure a GitHub package repository to resolve a specific dependency.
+
+```toml
+[[repository.github]]
+id = "github_1" # This id is used when pushing/ pulling balas
+url = "https://maven.pkg.github.com/ballerina-platform/ballerina-release"
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:177]:
```suggestion
url = "https://maven.pkg.github.com/jackson12/mypackagesrepo"
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 15]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+### Publish the Ballerina library package to a custom repository
+
+1. Generate the Ballerina archive. 
+   ```
+   $ bal pack
+   ```
+
+2. Publish to the custom repository.
+   ```
+   $ bal push --repository <repo-id>
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:191]:
```suggestion
   $ bal push --repository <repository-id>
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 16]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+   ```
+
+   If you already have the path of Ballerina archive, then,
+
+    ```
+    $ bal push --repository <repo-id> <path-to-bala-archive>
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:197]:
```suggestion
    $ bal push --repository <repository-id> <path-to-bala-archive>
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 17]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+    ```
+
+### Pull the Ballerina library package from a custom repository
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:200]:
```suggestion
### Pull a Ballerina library package from a custom repository
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 18]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+
+1. Ballerina package from the custom remote repository can be pulled using the following command.
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:202]:
```suggestion
1.Use the following command to pull a Ballerina package from a custom repository.  
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 19]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+
+    ```
+    $ bal push --repository <repo-id> <path-to-bala-archive>
+    ```
+
+### Pull the Ballerina library package from a custom repository
+
+1. Ballerina package from the custom remote repository can be pulled using the following command.  
+    ```
+   $ bal pull <org>/<package-name>:<version> --repository <repo-id>
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:204]:
```suggestion
   $ bal pull <package-org>/<package-name>:<version> --repository <repository-id>
```

The `package-org` is equivalent to the `group-id` whereas the `package-name` is equivalent to the `artifact-id`.

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 20]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+    ```
+
+### Use a dependency from a custom repository with build
+
+1. Similar to `local` repository, custom repository packages can be specified in the `Ballerina.toml` file and utilised in a similar manner.
+
+    ```toml
+    [[dependency]]
+    org = "boor"
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:213]:
```suggestion
    org = "jackson"
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 21]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+    name = "encrypter"
+    version = "2.1.1"
+    repository = "github_1" # Repository id is from above Settings.toml
```
Comment (azinneera) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:216]:
```suggestion
    repository = "github_1" # must match a repository id configured in the Settings.toml
```

**Review by azinneera — APPROVED** [review]

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 22]
```diff
@@ -29,6 +29,10 @@ The Ballerina Central is a remote repository and creates a local file system cac
 The local repository is also a file system repository, which will be created in the `<USER_HOME>` location. The repository location is `<USER_HOME>/.ballerina/repositories/local/bala`. 
 For more information, see [Use dependencies from the local repository](/learn/manage-dependencies/#use-dependencies-from-the-local-repository).
 
+**Custom repositories**
+
+Ballerina supports one or more custom remote repositories which can be configured in the `<USER_HOME>/.ballerina/Settings.toml`. A local filesystem cache is maintained per repository at `<USER_HOME>/.ballerina/repositories/<REPOSITORY_ID>/bala`. Ballerina queries the remote repository only if the specified dependency version is not present in its local cache. For more information, see [Use dependencies from a custom repository](/learn/manage-dependencies/#use-dependencies-from-a-custom-repository).
```
Comment (praneesha) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:34]:
```suggestion
Ballerina supports one or more custom remote repositories, which can be configured in the `<USER_HOME>/.ballerina/Settings.toml` file. A local filesystem cache is maintained per repository at `<USER_HOME>/.ballerina/repositories/<REPOSITORY_ID>/bala`. Ballerina queries the remote repository only if the specified dependency version is not present in its local cache. For more information, see [Use dependencies from a custom repository](/learn/manage-dependencies/#use-dependencies-from-a-custom-repository).
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 23]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
 However, the compiler gives priority to the latest version if a new patch version is found in distribution or Ballerina Central repositories.
```
Comment (praneesha) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:157]:
```suggestion
However, the compiler gives priority to the latest version if a new patch version is found in the distribution or Ballerina Central repositories.
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 24]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+   ```
+   $ bal pack
+   ```
+
+2. Publish to the custom repository.
+   ```
+   $ bal push --repository <repository-id>
+   ```
+
+   If you already have the path of Ballerina archive, then,
```
Comment (praneesha) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:194]:
```suggestion
   If you already have the path of the Ballerina archive, execute the below command.
```
Shall we make this a note?

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 25]
```diff
@@ -153,6 +157,65 @@ Ballerina considers the version specified in the Ballerina.toml as the minimum r
...
+
+    ```
+    $ bal push --repository <repository-id> <path-to-bala-archive>
+    ```
+
+### Pull a Ballerina library package from a custom repository
```
[the remark this answers is not in the document]
Comment (Thevakumar-Luheerathan) [reply swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:200]: This suggestion was given by @azinneera. AFAIK It emphasis the meaning of library

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 26]
```diff
@@ -14,7 +14,7 @@ A package can depend on other packages that are available in Ballerina repositor
 * The distribution repository
 * The Ballerina Central repository
 
-It also supports a third repository named the `local repository`. It temporarily overrides dependencies, which is useful for the package development and bug fixing phases.
+It also supports a third repository named the `local repository`. It temporarily overrides dependencies, which is useful for the package development and bug fixing phases. Additionally, a predefined set of custom package repositories are also supported, which are useful to bring third-party repositories into dependency management.
```
Comment (praneesha) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:17]:
```suggestion
It also supports a third repository named the `local repository`, and temporarily overrides dependencies, which is useful for the package-development and bug-fixing phases. Additionally, a predefined set of custom package repositories are also supported, which are useful to bring third-party repositories into dependency management.
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 27]
```diff
@@ -150,9 +154,68 @@ The local repository is useful to test a package in the development phase or to
 
 Once you complete the above steps, the dependency will be picked from the local repository when building the package.
 Ballerina considers the version specified in the Ballerina.toml as the minimum required version and uses the local repository to resolve the dependency.
```
Comment (praneesha) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:156]:
```suggestion
Ballerina considers the version specified in the `Ballerina.toml` file as the minimum required version and uses the local repository to resolve the dependency.
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 28]
```diff
@@ -150,9 +154,68 @@ The local repository is useful to test a package in the development phase or to
...
-However, the compiler gives priority to the latest version if a new patch version is found in distribution or Ballerina Central repositories.
+However, the compiler gives priority to the latest version if a new patch version is found in the distribution or Ballerina Central repositories.
 At this point, the compiler resolves the latest version and ignores the dependency version in the local repository.
 
+
+## Use dependencies from a custom repository
+
+Ballerina supports Maven repositories such as [Nexus](https://www.sonatype.com/products/sonatype-nexus-repository), [Artifactory](https://jfrog.com/artifactory/) and [Github packages](https://github.com/features/packages) to be set up as custom repositories. You can configure the custom repository in the `<USER_HOME>/.ballerina/Settings.toml` file to integrate it to the package resolution.
```
Comment (praneesha) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:163]:
```suggestion
Ballerina supports Maven repositories such as [Nexus](https://www.sonatype.com/products/sonatype-nexus-repository), [Artifactory](https://jfrog.com/artifactory/), and [Github packages](https://github.com/features/packages) to be set up as custom repositories. You can configure the custom repository in the `<USER_HOME>/.ballerina/Settings.toml` file to integrate it into the package resolution.
```

File: swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md [hunk 29]
```diff
@@ -150,9 +154,68 @@ The local repository is useful to test a package in the development phase or to
...
+
+    If you already have the path of the Ballerina archive, execute the below command.
+
+    ```
+    $ bal push --repository <repository-id> <path-to-bala-archive>
+    ```
+
+### Pull a Ballerina library package from a custom repository
+
+1. Use the following command to pull a Ballerina package from a custom repository.
```
Comment (praneesha) [inline swan-lake/learn-the-platform/source-code-dependencies/manage-dependencies.md:202]:
```suggestion
1. Execute the following command to pull a Ballerina package from a custom repository.
```
Be consistent with the usage of the word when they can be used interchangeably.

**Review by praneesha — APPROVED** [review]

**Pull request merged**